src/main/java/org/folio/rs/service/ReturnItemService.java

```diff
@@ -0,0 +1,98 @@
+import org.folio.rs.domain.dto.Request;
+import org.folio.rs.domain.dto.ReturnItemResponse;
+import org.folio.rs.domain.dto.User;
+import org.folio.rs.domain.entity.RetrievalQueueRecord;
+import org.folio.rs.error.ItemReturnException;
+import org.folio.rs.repository.RetrievalQueueRepository;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+import java.util.Collections;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+
+@Service
+@RequiredArgsConstructor
+@Log4j2
+public class ReturnItemService {
+
+  private static final String BARCODE_QUERY_PROPERTY = "barcode==";
+  private static final String USER_ID_QUERY_PROPERTY = "id==";
+
+  private final InventoryClient inventoryClient;
+  private final CirculationClient circulationClient;
+  private final UsersClient usersClient;
+  private final RetrievalQueueRepository retrievalQueueRepository;
+  private final CheckInItemService checkInItemService;
+
+  public ReturnItemResponse returnItem(String remoteStorageConfigurationId, CheckInItem checkInItem) {
+    log.info("Start return for item with barcode " + checkInItem.getItemBarcode());
+    var itemReturnResponse = new ReturnItemResponse();
+    var items = inventoryClient.getItemsByQuery(BARCODE_QUERY_PROPERTY + checkInItem.getItemBarcode());
+    if (items.isEmpty()) {
+      throw new ItemReturnException("Item does not exist for barcode " + checkInItem.getItemBarcode());
+    }
+    var item = items.getResult().get(0);
+    var requests = circulationClient.getItemRequests(item.getId());
+    if (!requests.isEmpty()) {
+      var holdRecallRequests = requests.getResult().stream()
+        .filter(request -> request.getRequestType() == Request.RequestType.HOLD
+          || request.getRequestType() == Request.RequestType.RECALL)
+        .collect(toList());
+      if (!holdRecallRequests.isEmpty()) {
+        itemReturnResponse.isHoldRecallRequestExist(true);
+        holdRecallRequests.stream()
+          .filter(itemRequest -> itemRequest.getPosition() == 1)
+          .findFirst().ifPresent(itemRequest-> {
+            var users = usersClient.getUsersByQuery(USER_ID_QUERY_PROPERTY + itemRequest.getRequesterId());
+            if (users.isEmpty()) {
+              throw new ItemReturnException("User does not exist for requester id " + itemRequest.getRequesterId());
+            }
+            retrievalQueueRepository.save(buildRetrievalRecord(itemRequest, item, users.getResult().get(0), remoteStorageConfigurationId));
+        });
+      }
+    }
+    checkInItemService.checkInItemByBarcode(remoteStorageConfigurationId, checkInItem);
+    log.info("Return success for item with barcode " + checkInItem.getItemBarcode());
+    return itemReturnResponse;
+  }
+
+  private RetrievalQueueRecord buildRetrievalRecord(Request itemRequest, Item item, User patron, String remoteStorageId) {
```

Comment:
Personally me prefer to use Spring Converter<A, B> for these purposes and create separate converter component.